### PR TITLE
Migrating to 0.4.0

### PR DESCRIPTION
**Pull request by joelgrus**
my plan is to link to this from the release notes.

let me know if I missed anything or explained anything poorly.

### CODE REVIEW

**Comment by joelgrus** [comment]
@nelson-liu also interested in your feedback

**Comment by nelson-liu** [comment]
thanks, this looks great!

File: tutorials/misc/migrating-to-0.4.0.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    for line in data_file:
+       instance = ...
+       # yield instances rather than collecting them in a list
+       yield instance
+```
+
+In addition, the base `DatasetReader` constructor now takes a `lazy: bool` parameter,
+which means that your subclass constructor should also take that parameter
+(unless you don't want to allow laziness, but why wouldn't you?)
+and explicitly pass it the superclass constructor:
```
Comment (matt-gardner) [inline tutorials/misc/migrating-to-0.4.0.md]: "pass it to"?

File: tutorials/misc/migrating-to-0.4.0.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+```python
+class MyDatasetReader(DatasetReader)
+    def __init__(self,
+                 # other arguments
+                 lazy: bool = False):
+        super().__init__(lazy)
+        # whatever other initialization you need
+```
+
+(For the reasoning behind this change, see the [Laziness tutorial](https://github.com/allenai/allennlp/blob/master/tutorials/getting_started/laziness.md).)
```
Comment (matt-gardner) [inline tutorials/misc/migrating-to-0.4.0.md]: I'd remove the parentheses here.

File: tutorials/misc/migrating-to-0.4.0.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+# CHANGES YOU ARE MUCH LESS LIKELY TO RUN INTO
+
+If you only ever use our the command line tools (`python -m allennlp.run ...`) to train / evaluate models,
```
Comment (matt-gardner) [inline tutorials/misc/migrating-to-0.4.0.md]: You could say something in here about writing model / dataset reader code ("use our command line tools" doesn't make it obvious that you also shouldn't have to worry about anything else if you only wrote `Models` or `DatasetReaders`).  Also, you have "use our the command line tools".

File: tutorials/misc/migrating-to-0.4.0.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+AllenNLP 0.4.0 gets rid of the `Dataset` class / abstraction.
+Previously, a `Dataset` was a wrapper for a concrete list of instances
+that contained logic for _indexing_ them with some vocabulary,
+and for converting them into `Tensor` s. And, as mentioned above,
+previously `DatasetReader.read()` returned such a dataset.
+
+In 0.4.0, `DatasetReader.read()` returns an `Iterable[Instance]`,
+which could be a list of instances or could produce them lazily.
+This means that the indexing and tensorization needs to happen elsewhere.
```
Comment (matt-gardner) [inline tutorials/misc/migrating-to-0.4.0.md]: A sentence saying where the logic moved would be nice.

File: tutorials/misc/migrating-to-0.4.0.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+## `Vocabulary`
+
+As `Dataset` no longer exists, we replaced `Vocabulary.from_dataset()`
+with `Vocabulary.from_instances()`, which accepts an `Iterable[Instance]`.
+In particulary, you'd most likely call this with the results of one or more calls
```
Comment (matt-gardner) [inline tutorials/misc/migrating-to-0.4.0.md]: "particulary"

File: tutorials/misc/migrating-to-0.4.0.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+to `DatasetReader.read()`.
+
+## `Batch`
+
+To handle tensorization,
+0.4.0 introduces the notion of a `Batch`,
+which is basically just a list of `Instance`s.
```
Comment (matt-gardner) [inline tutorials/misc/migrating-to-0.4.0.md]: Above you have `Tensor` s (with a space) and here you have it with no space.  Not sure which is right.

File: tutorials/misc/migrating-to-0.4.0.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+are loaded into memory (and then discarded) as needed.
+
+Accordingly, we moved the indexing functionality into
+`DataIterator.index_with()`, which hangs onto the
+`Vocabulary` you provide it and indexes each `Instance`
+as it's iterated over.
+
+Furthermore, each `Instance` now knows whether it's been indexed,
+so in the eager case (when all instances stay in memory),
+the indexing only happens in the first iteration.
```
Comment (matt-gardner) [inline tutorials/misc/migrating-to-0.4.0.md]: Something like "This means that your first epoch will bit a little bit slower than all other epochs" would be good, I think.

**Pull request merged**